WooCommerce : Add support for fetching custom order stats

anitaa1990 opened the pull request:
Fixes #1007 

This PR adds the following:

* Provides option to pass date range when fetching order/revenue stats
* Calculates quantity (how far back to go), based on the date range and the granularity before passing it to server api. If no date range is passed, default/static values are used.
* Parses and caches the response in the db - including the quantity and date.
*  Unit tests and mock tests included. (Added in `WCStatsStoreTest`)
* Also includes 2 new buttons on the **Woo** screen of the **example** app: **FETCH CUSTOM REVENUE STATS** and **FETCH CUSTOM REVENUE STATS (FORCED)**

<img src="https://user-images.githubusercontent.com/22608780/52257088-40e5ef00-293f-11e9-8bd1-20c56e8400a2.png" width="300"/>

<img src="https://user-images.githubusercontent.com/22608780/52913540-1738aa80-32e5-11e9-87f5-58f3c6e2b81d.png" width="300"/>

## Testing Notes
*  This feature includes an alteration to existing db table: `WCOrderStatsModel` - App upgrade testing to be done as well.

**Note** Changes need to be added to the front end as well to utilise this feature in WCAndroid.

Commit:
Add utils to calculate quantity based on start date and end date

Commit:
Add utils method to format start date before fetching data from api

Commit:
Modified utils method to calculate quantity to support previously missed scenarios

Commit:
Moved utils method to calculate quantity to WCStatsStore class

Commit:
Modified method to format date based on site's timezone

Commit:
Modified method to fetch order stats - to include start date and end date

Commit:
Merge branch 'develop' into feature/support_custom_order_stats

Commit:
Added logic to include date, quantity and customField when querying and inserting data to local cache

Commit:
Added logic to include query based on site and granularity only if quantity or date is null

Commit:
Modified : current logic to fetch order stats - to support custom stats

Commit:
Added logic to add columns to db table when app is upgraded to current version

Commit:
8. Added UI option in example module to test WCStatsStore logic for custom stats

Commit:
Formatted code to fix Travis CI - Build #6774 failed status

Commit:
Formatted code to fix Travis CI - Build #6775 failed status

AmandaRiu reviewed (commented):
@anitaa1990  very nice work. I left some small lower priority comments in the code, but the overall logic looks great. I played around with the buttons in the example app, that looks good. Could you add some logic to remember the last selection in the custom stats dialog? Not a big deal, will just make testing easier.

I ran out of time to really dig into the code as much as I would like to, but I did review the code changes and play with the example app and on this first pass it looks great. Nice attention to detail and your code is very clean. My only suggestion is to make sure you run the `a8c style` checkstyle config on the project after you make changes, as well as `./gradlew ktlint` from the command line. I'll do a more in-depth review tomorrow and submit more feedback once the small suggested fixes have been submitted.

anitaa1990 commented on fluxc/src/main/java/org/wordpress/android/fluxc/utils/SiteUtils.java:
```diff
@@ -98,4 +124,169 @@
+     * returns a {@link Date} instance
+     * based on {@param pattern} and {@param dateString}
+     */
+    public static Date getDateFromString(String dateString) {
```
Ahh your're right! Will do that!

anitaa1990 replied:
Moved the methods to a separate class in this [commit](https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/1103/commits/ef8d9e3b20b5ccca8eed32715152c5f024a754b3)

anitaa1990 commented on fluxc/src/main/java/org/wordpress/android/fluxc/utils/SiteUtils.java:
```diff
@@ -2,16 +2,42 @@
 public class SiteUtils {
+    private static final String DATE_FORMAT_DEFAULT = "yyyy-MM-dd";
+
+
```
Yup:) Actually, noticed that when I created PR, the CI build was failing because of this so added [commit](https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/1103/commits/07269b35d16c49dc6c3d579145477b066fdadcb0) to fix it

anitaa1990 commented on fluxc/src/main/java/org/wordpress/android/fluxc/utils/SiteUtils.java:
```diff
@@ -2,16 +2,42 @@
+    /**
+     * Given a {@link SiteModel} and a {@link String} compatible with {@link SimpleDateFormat} and a {@param dateString}
+     * returns a formatted date that accounts for the site's timezone setting.
+     *
```
Yup:) Actually, noticed that when I created PR, the CI build was failing because of this so added [commit](https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/1103/commits/07269b35d16c49dc6c3d579145477b066fdadcb0) to fix it

Commit:
Move non site related code from SiteUtils.java to DateUtils class

Commit:
Replace date in sql table with start date and end date

Commit:
Added method to get start date and end date from sql

Commit:
Added method to get start date and end date from sql

Commit:
Added option to store startDate in sql

Commit:
Added option to store endDate and startDate in sql

anitaa1990 commented:
@AmandaRiu as per our discussion, I added the startDate to sqlite. I also added the `endDate` field as well. Even though we are storing the end date field (this was formatted based on the granularity - so if WEEKS were selected, then the date would be stored in yyyy-'W'ww format). Since this would not accurately display the end date the user has selected, I had added this field as well.

Please review :) 

I also want to add that: I think there have been some changes to the `WellSqlConfig.java` class. When I started development, the db version was 52. I think it has changed to 54. So when merging with the develop branch, we would need to modify the class to update our changes from this branch as well.

Commit:
modified default value of startDate and endDate to null

Commit:
Resolve conflicts and merge with develop branch

anitaa1990 commented:
@AmandaRiu merged with the latest changes in develop branch and resolved the conflicts. Please review :)

Commit:
Modified fetch stats method to set start and end date to null if no value is passed

Commit:
Added logic to StatsGranularity to get value from string

Commit:
Optimised previous commit to fix ktlint errors

AmandaRiu reviewed (changes requested):
@anitaa1990 This is looking great! I do have several code comments in addition to items noted below. In the code comments, "np" is our shortened code for "nitpick" which basically means it's a really low-priority comment and more of a suggestion than anything else. Typically if a reviewer adds a nitpick, we implement the suggestion unless we feel very strongly that it shouldn't be implemented. 

Here are some other things I found during my review:
- The date selector dialog remembers the selected dates, but once you select a date in the past, you can't go forward on the calendar. Also, when you open the calendar again, the selected date is off by a month.
<img src="https://user-images.githubusercontent.com/5810477/52678784-e16c7e00-2f00-11e9-8a1e-9051fc9932db.gif" width="300"/>

- Date selector loses selection if left open while switching apps, and then doesn't register changes (probably because the listener will be gone) -- to test - make sure "DON'T KEEP ACTIVITIES" is toggled on in device developer options. 
<img src="https://user-images.githubusercontent.com/5810477/52678819-f0ebc700-2f00-11e9-8c4f-6e818af01380.gif" width="300"/>

- 30 days comparison w/ standard stats: the end date ends up one day different (selecting a start date of 1/14 and 30 days):
	- standard: 2/12
	- custom: 2/11
- custom ranges for fetching visitor stats is not included yet. This will be needed for the card in the UI.

Commit:
Added fixes to date selector dialog in example app

Commit:
Modified code in WCStatsStore fetchStats method to increase readability

Commit:
Modified logic when formatting end date to include timezone

Commit:
Convert DateUtils java class to kotlin and added test cases

Commit:
Remove redundant methods in WcStatsSqlUtils

Commit:
Modified custom_stats_dialog.xml to include single line space between elements

Commit:
Modified raw stats method in WCStatsSqlUtils file to include custom stats field in query

Commit:
Fixed unit test cases for DateUtils

Commit:
Added isCustomField to OnWCStatsChanged to differentiate custom and default stats

Commit:
Added logic to include startDate and endDate for fetching visitor stats

Commit:
Modified logic to fix date selector selection loss while switching apps

anitaa1990 commented on example/src/main/java/org/wordpress/android/fluxc/example/WooCommerceFragment.kt:
```diff
@@ -427,12 +454,18 @@ class WooCommerceFragment : Fragment() {
+                        prependToLog("Fetched stats for " + statsMap.size + " " +
+                                event.granularity.toString().toLowerCase() + " from " + site.name +
+                                " with quantity " + event.quantity + " and date " + event.date)
+                    } ?: run {
```
I added a new field to the `OnWCStatsChanged` class, to differentiate between custom stats and default stats. So modified logic based on this.

anitaa1990 commented on plugins/woocommerce/src/main/kotlin/org/wordpress/android/fluxc/store/WCStatsStore.kt:
```diff
@@ -231,10 +255,55 @@ class WCStatsStore @Inject constructor(
+        val quantity = getQuantityForGranularity(payload.site, payload.granularity, payload.startDate, payload.endDate)
         wcOrderStatsClient.fetchStats(payload.site, OrderStatsApiUnit.fromStatsGranularity(payload.granularity),
-                getFormattedDate(payload.site, payload.granularity), quantity, payload.forced)
+                getFormattedDate(payload.site, payload.granularity, payload.endDate),
```
:) got it! Modified this and all corresponding methods to increase readability

anitaa1990 commented on plugins/woocommerce/src/main/kotlin/org/wordpress/android/fluxc/persistence/WCStatsSqlUtils.kt:
```diff
@@ -44,4 +44,76 @@ object WCStatsSqlUtils {
                 .asModel.firstOrNull()
     }
+
+    fun insertOrUpdateOrderStats(stats: WCOrderStatsModel): Int {
```
Yeah, makes sense to remove the new method and add the updated logic to the old method. Also `event.date` is never null since the date from the server api is never null. So added a new field  `event.isCustomField` to differentiate between default and custom stats

anitaa1990 commented on plugins/woocommerce/src/main/kotlin/org/wordpress/android/fluxc/persistence/WCStatsSqlUtils.kt:
```diff
@@ -44,4 +44,76 @@ object WCStatsSqlUtils {
+        date: String? = null
+    ): WCOrderStatsModel? {
+        if (quantity.isNullOrEmpty() || date.isNullOrEmpty())
+            return getRawStatsForSiteAndUnit(site, unit)
```
Based on this logic, I've modified all the queries to include `IS_CUSTOM` - since as you said if custom stats has the same date range and unit as default stat, then it would unintentionally return custom stat instead of default stat.

anitaa1990 commented on plugins/woocommerce/src/main/kotlin/org/wordpress/android/fluxc/utils/DateUtils.java:
```diff
@@ -0,0 +1,199 @@
+
+        return cal1;
+    }
+
```
👍 Modified this as well!

anitaa1990 commented on plugins/woocommerce/src/main/kotlin/org/wordpress/android/fluxc/utils/DateUtils.java:
```diff
@@ -0,0 +1,199 @@
+     * Given a {@link Calendar} instance for startDate and endDate,
+     * returns a quantity that is calculated based on [StatsGranularity.MONTHS]
+     */
+    public static long getQuantityInMonths(@NonNull Calendar c1,
```
Makes sense! Converted this class to kotlin and added some more test cases for the methods

anitaa1990 commented on plugins/woocommerce/src/main/kotlin/org/wordpress/android/fluxc/utils/DateUtils.java:
```diff
@@ -0,0 +1,199 @@
+     * returns a formatted date that accounts for the site's timezone setting.
+     *
+     */
+    public static @NonNull String getDateTimeForSite(@NonNull SiteModel site,
```
👍

anitaa1990 commented on example/src/main/res/layout/dialog_custom_stats.xml:
```diff
@@ -0,0 +1,104 @@
+        android:textAppearance="@style/Base.TextAppearance.Widget.AppCompat.Toolbar.Subtitle"
+        app:layout_constraintStart_toStartOf="parent"
+        app:layout_constraintTop_toBottomOf="@+id/stats_dialog_title"/>
+
```
👍

anitaa1990 commented on plugins/woocommerce/src/main/kotlin/org/wordpress/android/fluxc/utils/DateUtils.java:
```diff
@@ -0,0 +1,199 @@
+import java.util.Date;
+import java.util.Locale;
+
+public class DateUtils {
```
👍 Modified it to kotlin and added test cases as well

Commit:
Removed redundant method from WCStatsStore

AmandaRiu reviewed (changes requested):
@anitaa1990 Nice work on this PR...it's almost ready! Just a few more items to address:
- I see you added support for custom visitor stats, but I don't see a button in the example app to test the custom visitor stats.
- The custom date range selector dialog does not scroll. This is needed when in landscape mode:
<img src="https://user-images.githubusercontent.com/5810477/52875451-62ef2680-3122-11e9-9398-0732a063643d.png" width="300"/>
- Date selector being off by a month is fixed, but I'm seeing some strange behavior when the selected date is the last day of the month:
<img src="https://user-images.githubusercontent.com/5810477/52875564-9d58c380-3122-11e9-9b27-668e3857e218.gif" width="300"/>
- Please also see code comments.

AmandaRiu commented:
@anitaa1990 I've created a matching branch in our FluxC repository and will merge this PR into that branch once you address my last review. This will make it easier for us to test going forward because once this is merged we can just use the commit hash in the WCAndroid code directly.

Commit:
Modified logic to format calendar date in sample app and added unit test cases

anitaa1990 commented on plugins/woocommerce/src/main/kotlin/org/wordpress/android/fluxc/utils/DateUtils.kt:
```diff
@@ -0,0 +1,214 @@
+        calendar.time = date
+        calendar.add(Calendar.HOUR_OF_DAY, currentDate.hours)
+        calendar.add(Calendar.MINUTE, currentDate.minutes)
+        calendar.add(Calendar.SECOND, currentDate.seconds)
```
Sure. I added a calendar instance instead to get the hours, minutes and seconds. You can checkout this [commit](https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/1103/commits/810349f9fd6879596369f7b39bb6ab025fd3ba96).  

The commit also includes the changes to the date utils class - this was originally in the sample app but since we use the same utils to display in both sample and woocommerce app, I migrated it to the DateUtils class. Fixed the strange behaviour of date selector as well - when last date of month was selected. I added test cases to test all dates from 2011 till present.

Commit:
Added scrollView to custom stats dialog to support landscape mode

Commit:
Added 2 buttons in example app to get custom visitor stats

anitaa1990 commented:
> I see you added support for custom visitor stats, but I don't see a button in the example app to test the custom visitor stats.
Hi @AmandaRiu , this has actually been on my TODO list. I have added 2 buttons * on the **Woo** screen of the **example** app to * **FETCH CUSTOM VISITOR STATS** and **FETCH CUSTOM VISITOR STATS (FORCED)**. There are a couple of clarifications to the implementation though:
* The date selection for visitor stats is not remembered. For order stats, the date selection is stored in FluxC. But for visitor stats, since this functionality is not needed in FluxC, I thought of adding the visitor date selection and granularity in SharedPreferences and displaying that.
* Unlike `FetchOrderStatsResponsePayload`, `FetchVisitorStatsResponsePayload` does not include the `startDate` and `endDate` selected by user. So when the `FETCH_VISITOR_STATS` event is fired, the sample app will not display the date selected by the user in the logs. But without this there is no differentiation between the default visitor stats and the custom stats. So I thought of including the `date`, `quantity` and `isCustomField` for `FETCH_VISITOR_STATS` as well. This feature will not be used at the moment in the WooCommerce app but in the example app, we would be able to differentiate between the custom and default visitor stats. 

> The custom date range selector dialog does not scroll. This is needed when in landscape mode
This is added in this [commit](https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/1103/commits/12ca7c88ad58bc86a1a88d57d349e6538db907bf)

> Date selector being off by a month is fixed, but I'm seeing some strange behavior when the selected date is the last day of the month
This [commit](https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/1103/commits/810349f9fd6879596369f7b39bb6ab025fd3ba96) includes the changes to the date utils class - this was originally in the sample app but since we use the same utils to display in both sample and woocommerce app, I migrated it to the DateUtils class. Fixed the strange behaviour of date selector as well - when last date of month was selected. I added test cases to test all dates from 2011 till present.

Commit:
Added logic to remember date selection for visitor stats in example app

Commit:
Added logic to pass date, quantity and isCustomField to visitor stats event

AmandaRiu commented on example/src/main/java/org/wordpress/android/fluxc/example/CustomStatsDialog.kt:
```diff
@@ -73,10 +74,10 @@ class CustomStatsDialog : DialogFragment() {
             forced = savedInstanceState.getBoolean("forced")
         } else {
-            val startDate = wcOrderStatsModel?.startDate ?: getCurrentDateString()
+            val startDate = wcOrderStatsModel?.startDate ?: DateUtils.getCurrentDateString()
```
It was a really good call moving these utilities into the main `DateUtils` class. 👍

AmandaRiu reviewed (approved):
@anitaa1990 This looks great! There are a few things I'll make notes on in a new PR after this is merged into the repo branch I had previously talked about, but overall, we're ready to move it under the FluxC umbrella!

AmandaRiu commented:
@anitaa1990 For the visitor stats code: I recommend removing anything that was added to the Payload specifically for the example qpp that wouldn't be used by the main app since it's just extra baggage. The visitor api calls all use caching at the HTTP level so if the parameters are the same, it will just automatically use the matching response. I also recommend removing the code for shared preferences. For the example app, just having the date ranges cached for the _active session_ is enough. Does that sound good? 

The rest all looks really good 👍 

You may need to create a new PR to merge into my `anitaa/support_custom_order_stats ` branch up on FluxC for these changes so I can merge them into our current branch since I don't believe you have access to edit that PR or branch directly (although you are welcome to try).

ParaskP7 commented:
👋 @anitaa1990 @AmandaRiu !

It's been 3,5 years already, I know... 😅 But, let me ask you this millisecond related question below, just in case you still remember the answer. 🤞 

As part of [this Detekt related issue](https://github.com/wordpress-mobile/WordPress-FluxC-Android/issues/2486) that I am currently working on, I am trying to resolve the `MagicNumber` related warnings. To that end, I stumbled upon this `1` warning, on the `plugins:woocommerce` module, that is related to this [DateUtils](https://github.com/wordpress-mobile/WordPress-FluxC-Android/blob/trunk/plugins/woocommerce/src/main/kotlin/org/wordpress/android/fluxc/utils/DateUtils.kt) class that you both worked on (see [here](https://github.com/wordpress-mobile/WordPress-FluxC-Android/blob/trunk/plugins/woocommerce/src/main/kotlin/org/wordpress/android/fluxc/utils/DateUtils.kt#L103)) and was wondering about it. Was there a reason you used the `59` instead of `999` for the millisecond value there? 🤔
